packages/insured-bridge-relayer/src/Relayer.ts

```diff
@@ -56,73 +64,165 @@ export class Relayer {
               message: "Not relaying deposit 😖",
-              l1Token,
+              realizedLpFeePct,
               relayableDeposit,
-              botL1TokenBalance,
             });
             break;
           case RelayType.Slow:
-            this.logger.info({
-              at: "Relayer",
-              message: "Slow relaying deposit 🐌",
-              l1Token,
+            this.logger.debug({
+              at: "InsuredBridgeRelayer#Relayer",
+              message: "Slow relaying deposit",
+              realizedLpFeePct,
               relayableDeposit,
-              botL1TokenBalance,
             });
-            await this.slowRelay(relayableDeposit);
+            await this.slowRelay(relayableDeposit.deposit, realizedLpFeePct);
             break;
 
-          case RelayType.Fast:
-            this.logger.info({
-              at: "Relayer",
-              message: "Fast relaying deposit 🏃‍♂️",
-              l1Token,
+          case RelayType.SpeedUp:
+            this.logger.debug({
+              at: "InsuredBridgeRelayer#Relayer",
+              message: "Speeding up existing relayed deposit",
+              realizedLpFeePct,
               relayableDeposit,
-              botL1TokenBalance,
             });
-            await this.fastRelay(relayableDeposit);
+            await this.speedUpRelay(relayableDeposit.deposit);
+            break;
+          case RelayType.Instant:
+            this.logger.debug({
+              at: "InsuredBridgeRelayer#Relayer",
+              message: "Instant relaying deposit",
+              realizedLpFeePct,
+              relayableDeposit,
+            });
+            await this.instantRelay(relayableDeposit.deposit);
             break;
         }
       }
     }
   }
 
-  private shouldRelay(deposit: Deposit, botTokenBalance: string): RelayType {
-    // shouldRelay(deposit) => RelayType { No, Slow, Fast }
-    // Checks bot balance.
-    // If bot balance is > transfer size, simulate the fast relay to determine the cost.
-    // If bot balance is > bond, simulate the slow relay to determine the cost.
-    // If the max profit across the two (or all that succeeded) is negative or 0, return No.
-    // Else return the one with the most profit.
-
-    // TODO: Add logic to decide if a relay action should happen
-    console.log(deposit, botTokenBalance);
-    return RelayType.Slow;
+  private async shouldRelay(deposit: Deposit, relayAbility: RelayAbility, realizedLpFeePct: BN): Promise<RelayType> {
+    const [l1TokenBalance, proposerBondPct] = await Promise.all([
+      getTokenBalance(this.l1Client.l1Web3, deposit.l1Token, this.account),
+      this.l1Client.getProposerBondPct(),
+    ]);
+    const relayTokenRequirement = this.getRelayTokenRequirement(deposit, proposerBondPct, realizedLpFeePct);
+
+    // There are three different kinds of profits that the bot can produce:
+    let slowProfit = toBN("0");
+    let speedUpProfit = toBN("0");
+    let instantProfit = toBN("0");
```

Comment:
How is instant profit diff from speedup profit?

Reply:
the diff is that: a) speedup is considering a slow relay that has not been sped up and b) instant is a deposit that has no relay on L1 at all. this bot needs to consider the 3 cases:
a) slow relay
b) speed up a slow relay and
c) instantly relay a deposit with no actions done on it.

these three capture these profits accordingly